plugins/newsql/newsql_evbuffer.c

```diff
@@ -520,37 +521,54 @@ static int newsql_read_evbuffer(struct sqlclntstate *clnt, void *b, int l, int n
     return have / l;
 }
 
-static int newsql_pack_hb(uint8_t *out, void *arg)
+static int newsql_pack_hb(struct sqlwriter *writer, void *arg)
 {
     struct sqlclntstate *clnt = arg;
     int state;
+    struct newsqlheader h = {0};
     if (is_pingpong(clnt))
         state = 1;
     else {
         state = (clnt->sqltick > clnt->sqltick_last_seen);
         clnt->sqltick_last_seen = clnt->sqltick;
     }
-    struct newsqlheader *h = (struct newsqlheader *)out;
-    memset(h, 0, sizeof(struct newsqlheader));
-    h->type = htonl(RESPONSE_HEADER__SQL_RESPONSE_HEARTBEAT);
-    h->state = htonl(state);
-    return 0;
+    h.type = htonl(RESPONSE_HEADER__SQL_RESPONSE_HEARTBEAT);
+    h.state = htonl(state);
+    return evbuffer_add(sql_wrbuf(writer), &h, sizeof(h));
 }
 
 struct newsql_pack_arg {
     struct newsqlheader *hdr;
     const CDB2SQLRESPONSE *resp;
 };
 
-static int newsql_pack(uint8_t *out, void *data)
+struct pb_evbuffer_writer {
+    ProtobufCBuffer base;
+    struct sqlwriter *sqlwriter;
+    int rc;
+};
+
+static void pb_evbuffer_write(ProtobufCBuffer *base, size_t len, const uint8_t *data)
+{
+    struct pb_evbuffer_writer *writer = (struct pb_evbuffer_writer *)base;
+    writer->rc |= sql_append_packed(writer->sqlwriter, data, len);
+}
+
+#define PB_EVBUFFER_WRITER_INIT(x) {.base = {.append = pb_evbuffer_write}, .sqlwriter = x, .rc = 0 }
+
+static int newsql_pack(struct sqlwriter *sqlwriter, void *data)
 {
     struct newsql_pack_arg *arg = data;
+    struct pb_evbuffer_writer writer = PB_EVBUFFER_WRITER_INIT(sqlwriter);
     if (arg->hdr) {
-        memcpy(out, arg->hdr, sizeof(struct newsqlheader));
-        out += sizeof(struct newsqlheader);
+        pb_evbuffer_write(&writer.base, sizeof(struct newsqlheader), (const uint8_t *)arg->hdr);
```

Comment:
These names are throwing me off! I parsed writer.base to mean sqlwriter and event_base. Any chance we can reword pb_evbuffer_writer and it's members.

Reply:
call it pb_writer (pb_base) perhaps. writer means sqlwriter everywhere..

Reply:
Sorry I missed your comment. I renamed it to `pb_evbuffer_appender`. I hope this is less confusing.